Add distributed ml reference architecture (#205)
crate/crate-howtos: docs/integrations/ml-dist.rst

======================================
Distributed Deep-Learning with CrateDB
======================================

Abstract
========
Using Deep-Learning algorithms for Machine Learning Use-Cases becomes
more and more common in the world of data science. A common library used

> for solving Deep-Learning problems is `TensorFlow`_, which will be used to
> predict the status of a water pump. CrateDB can easily be integrated into

Reviewer: In the Abstract section, we usually try to explain how to integrate CrateDB with the tools without mentioning any use cases, in this case the water pump. Maybe something like,

```suggestion
for solving Deep-Learning problems is `TensorFlow`_, can be used to generate
predictions based on various machine data features. CrateDB can easily be integrated into
```

> an ML-Application as a data source, as well as a data store for the resulting

Reviewer: ```suggestion
a Machine Learning application as a data source, as well as a data store for the resulting
```

predictions. The connection to CrateDB can be established using the `Psycopg2`_

> and `Pandas`_ library. To prepare the data by for example labeling it, we will
> be using the library `Sklearn`_.

Reviewer: ```suggestion
and `Pandas`_ library. Preparing the data can be done with `Sklearn`_.
```

Implementation
==============

Set Up
------

> For this distributed ML implementation, we will be using a `Kaggle`_ data set

Reviewer: ```suggestion
For this distributed machine learning implementation, we will be using a `Kaggle`_ data set
```

ML should be expanded to "machine learning"

called `pump_sensor_data`_.

> The data set contains data gathered from 52 sensors, with not much other detail
> to them. These data rows come with a timestamp and a machine-status attached to

Reviewer: ```suggestion
The data set contains data gathered from 52 sensors. 
These data rows come with a timestamp and a machine-status attached to
```

> them. There are 3 unique states the machine can have, *normal*, *broken*,
> *recovering*. The data is stored in the CrateDB table ``ml.pump_sensors``.

Reviewer: I think a list here would be better. Something like:

```suggestion
them. There are 3 unique states the machine can have:
- ``normal``
- ``broken``
- ``recovering``. 

The data is stored in the CrateDB table ``ml.pump_sensors``.
```

Using this sensor data we want to predict if the machine will fail in a
specified time window in the future.

Using this data with TensorFlow, we want to accomplish the following:

1. Retrieve data from CrateDB.
2. Show extracts of how the data looks.
3. Prepare the data and build a machine learning model that will predict
   if the machine will fail in a specified time window in the future.
4. Save the previously built model on AWS S3.
5. Retrieve the model from AWS S3, predict sample data and insert the results
   into the CrateDB table ``ml.pump_sensors_predictions``.

Prerequisites
-------------

- `A running CrateDB cluster`_
- `A AWS S3 storage bucket`_

> - A AWS config file at ``~/.aws/config`` containing:

Reviewer: Could you explain why this is needed? I assume for the boto3 library, but it would be worthwhile to say this here.

.. code-block:: python

    [default]
    aws_access_key_id=<key>
    aws_secret_access_key=<access_key>

- `Python`_
    - The `tensorflow`_ library
    - The `numpy`_ library
    - The `sklearn`_ library
    - The `pandas`_ library
    - The `psycopg2`_ library
    - The `boto3`_ library
    - The `joblib`_ library
    - The `matplotlib`_ library

CrateDB
-------

> In this first step we need to make sure that we create the table that holds the pump_sensor

Reviewer: ```suggestion
In this first step we need to make sure that we create the table that holds the pump sensor
```

> data set as well as the table that will be saving the predictions:

Reviewer: ```suggestion
data set as well as the table that will be storing the predictions:
```

       "sensor_18" REAL,
       "sensor_19" REAL,
       "sensor_20" REAL,
       "sensor_21" REAL,
       "sensor_22" REAL,
       "sensor_23" REAL,
       "sensor_24" REAL,
       "sensor_25" REAL,
       "sensor_26" REAL,
       "sensor_27" REAL,
       "sensor_28" REAL,
       "sensor_29" REAL,
       "sensor_30" REAL,
       "sensor_31" REAL,
       "sensor_32" REAL,
       "sensor_33" REAL,
       "sensor_34" REAL,
       "sensor_35" REAL,
       "sensor_36" REAL,
       "sensor_37" REAL,
       "sensor_38" REAL,
       "sensor_39" REAL,
       "sensor_40" REAL,
       "sensor_41" REAL,
       "sensor_42" REAL,
       "sensor_43" REAL,
       "sensor_44" REAL,
       "sensor_45" REAL,
       "sensor_46" REAL,
       "sensor_47" REAL,
       "sensor_48" REAL,
       "sensor_49" REAL,
       "sensor_50" REAL,
       "sensor_51" REAL,
       "machine_status" TEXT
    );

    CREATE TABLE "ml"."pump_sensors_predictions"
    (
       "prediction" REAL,
       "ts" TIMESTAMP WITH TIME ZONE,
       "model_name" TEXT
    );

After creating the tables we will start with looking at our data before we will
insert data to our table, since the data might not be in a state where it can

> be easily inserted into our CrateDB instance.

Reviewer: I would rephrase this to something like:

Before we insert any data into the ``pump_sensors`` table, we should first inspect it in case it needs processing before it can be successfully inserted.

Assumption
----------

The python code is considered to all be in a single ``__main__`` function
if nothing else was explicitly stated. Imports should all be at top of the
file, while the code snippets follow in order of each other from the top

> to the bottom of the file.

Reviewer: I think it's ok to remove this assumption section - the following could be done within a single python file or in the repl.

Examining The Data
------------------

We will begin with loading in our data to a pandas dataframe, so we can have

> a look at it and make adjustments accordingly.

Reviewer: I think this data examination/cleansing section is probably not necessary. We could maybe say in the prerequisites that we have some sample data that has been cleaned up for processing, since this section doesnt really involve doing anything with CrateDB.

    4   4  2018-04-01 00:04:00   2.445718   47.13541    53.2118  46.397568   636.4583   76.58897   13.35359   16.21094  ...   41.40625  38.773150   65.10416   51.79398  38.773150   158.2755   66.55093   242.1875   201.3889          NORMAL

    [5 rows x 55 columns]

    # We can see that the first column has no name
    data.rename(columns={data.columns[0]: "id"}, inplace=True)

    # Now let's see if the dataset has any missing values
    data.isna().sum().sum()

    [1] 344574

    # Ok so we have 344574 missing values, let's fill those up
    missing_values = -1.0
    data = data.fillna(missing_values)

Now that we have filled the missing values of the dataset, we can insert the
data to our CrateDB instance.

.. code-block:: python

    data.to_csv("sensor_new.csv", index=False)

    conn_str = "host=localhost user=crate port=5432"

    # This should only be run once, as it will duplicate the data in the table otherwise.
    with connect(conn_str) as conn:
        with conn.cursor() as cur:
            query = "COPY ml.pump_sensors FROM 'file:///path/to/sensor_new.csv'"
            cur.execute(query)

Training A Model
----------------

> First, we will load the data into our python context with ``Psycopg2`` and

Reviewer: ```suggestion
First, we will load the data into our Python context with ``Psycopg2`` and
```

``Pandas``

> .. code-block:: python
> 
>     import pandas.io.aql as sqlio
>     from psycopg2 import connect
> 
> 
>     with connect(conn_str) as conn:
>         query = "SELECT * FROM ml.pump_sensors;"
>         df = sqlio.read_sql_query(query, conn)

Reviewer: In this code block, and following code blocks, could you add a brief inline comment above each block of code about what is happening/what you're doing? Here it is simple enough, but later when you start doing stuff with sklearn/tf it can be a little difficult to follow. something like:

```suggestion
.. code-block:: python

    import pandas.io.aql as sqlio
    from psycopg2 import connect

    
    # Connecting to CrateDB and retrieving the pump sensor data.
    with connect(conn_str) as conn:
        query = "SELECT * FROM ml.pump_sensors;"
        df = sqlio.read_sql_query(query, conn)
```

> The data was successfully loaded for the application and we can now start

Reviewer: ```suggestion
Once the data has been loaded into a dataframe, we can now start
```

> to prepare the data so we can train a `Multilayer Perceptron`_ (MLP) to
> classify the machine-status in a given time window.

Reviewer: ```suggestion
to prepare the data. This is so we can train a `Multilayer Perceptron`_ (MLP) to
classify the machine status within a given time window.
```

> We will make an assumption and combine the two labels ``RECOVERING``
> and ``BROKEN`` and combine them as a single label, so we can either say
> the machine is in the state ``NORMAL`` or ``BROKEN``. Then we will binarize

Reviewer: ```suggestion
We will make an assumption and combine the two labels ``RECOVERING``
and ``BROKEN`` into a single label: ``BROKEN``. With this, we can either say
the machine is in the state ``NORMAL`` or ``BROKEN``. Then we will binarize
```

the labels so we can use them in the MLP. Furthermore we need to extract
the sensor columns and split the data into train and test data frames while
normalizing their values between (0,1).

.. code-block:: python

    from sklearn.preprocessing import MinMaxScaler, label_binarize

    df["machine_status"] = df["machine_status"].replace("RECOVERING", "BROKEN")
    df["machine_status"] = label_binarize(df["machine_status"], classes=["BROKEN", "NORMAL"])

    min_max_scaler = MinMaxScaler(feature_range=(0, 1))

    test_split = 70
    df_size = len(df.index)
    split = int((df_size / 100) * test_split)

    df_train = df.iloc[0:split, :]
    train_data_cols = df_train.iloc[:, 2:54]
    norm_train_data_cols = min_max_scaler.fit_transform(train_data_cols)

    df_test = df.iloc[split:len(df.index), :]
    test_data_cols = df_test.iloc[:, 2:54]
    norm_test_data_cols = min_max_scaler.transform(test_data_cols)

The next step would be to define our time steps for the data. Let's say we take
an input time of 60 minutes and try to predict if the machine will fail in the next

> 12 hours or 720 minutes. After defining these we will create data pairs for training

Reviewer: ```suggestion
12 hours. After defining these we will create data pairs for training
```

I think it's okay to leave that implicit :wink:

the MLP model that fit into the time steps.

.. code-block:: python

    # Add this function over the main function.
    def add_labels(start, stop, dataset):
        for step in range(start, stop):
            if dataset["machine_status"].iloc[step] == 0:
                return 0
        return 1

    timestep = 60
    output_timestep = 720

    train_pairs = []

    saved_step = 0
    for i in df_train.index:
        row = saved_step
        if row + timestep + output_timestep < len(df_train.index):
            data_window = norm_train_data_cols[row:row + timestep]
            data_vector = data_window.flatten()
            state = add_labels(row + timestep, row + timestep + output_timestep, df_train)
            train_pairs.append((data_vector, state))
            saved_step = row + timestep
        else:
            break

> Now that we have the data pairs we can start creating our x/y training data frames.

Reviewer: Could you briefly explain what this means? Perhaps say to train the data, we need to split it into 2 sets: x and y. Set x will be used to train the model, and set y will be used to verify the model's accuracy. (Assuming that I understand this correctly :sweat_smile:)

.. code-block:: python

    import numpy as np

    input_len = train_pairs[0][0].shape[0]

    x_train = np.zeros((len(train_pairs), input_len))
    y_train = np.zeros((len(train_pairs),))

    for index in range(0, len(train_pairs)):
        (x, y) = train_pairs[index]
        x_train[index, 0:input_len] = x
        y_train[index] = y

After creating our training data frames we can now continue with creating the MLP model.
We are going to use 3 layers, with the `ReLU`_ activation function for the first two

> layers. If the model should get worse while going through the epochs we will stop the
> training with the ``EarlyStopping`` callback.

Reviewer: ```suggestion
layers. Prevention of the model degrading over successive epochs is done by stopping
training using the ``EarlyStopping`` callback.
```

.. code-block:: python

    import tensorflow as f
    import matplotlib.pyplot as plt
    from tensorflow.python.keras.callbacks import EarlyStopping
    from tensorflow.keras import layers

    mlp_model = tf.keras.Sequential()

    mlp_model.add(layers.Dense(200, activation="relu", input_shape=(input_len,)))
    mlp_model.add(layers.Dense(40, activation="relu"))
    mlp_model.add(layers.Dense(1))

    mlp_model.compile(optimizer="adam",
                      loss=tf.keras.losses.MeanSquaredError(),
                      metrics=["accuracy"])

    es = EarlyStopping(monitor="loss", mode="auto", verbose=1, patience=50)
    hist = mlp_model.fit(x_train, y_train, epochs=500, callbacks=[es])

    plt.plot(hist.history["loss"])
    plt.show()

> .. figure:: mlp_model_train_loss.png

Reviewer: Could you explain what the the pattern in this graph signifies briefly?

   :align: left

Now that we have our MLP model, we can start preparing the test data to verify
the accuracy of the model.

.. code-block:: python

    test_pairs = []

    test_saved_step = 0
    for i in df_test.index:
        row = test_saved_step
        if row + timestep + output_timestep < len(df_test.index):
            data_window = norm_test_data_cols[row:row + timestep]
            data_vector = data_window.flatten()
            state = add_labels(row + timestep, row + timestep + output_timestep, df_test)
            test_pairs.append((data_vector, state))
            test_saved_step = row + timestep
        else:
            break

    x_test = np.zeros((len(test_pairs), input_len))
    y_test = np.zeros((len(test_pairs),))

    for index in range(0, len(test_pairs)):
        (x, y) = test_pairs[index]
        x_test[index, 0:input_len] = x
        y_test[index] = y

> The data is prepared in the same way as the training data was prepared, and now we can
> predict the test data and look at the accuracy.

Reviewer: ```suggestion
This data is prepared in the same way as the training data. It can now be used with the model
to generate predictions and to estimate the model's accuracy.
```

.. code-block:: python

    predictions = mlp_model.predict(x_test)
    predictions = predictions.flatten()

    _, train_acc = mlp_model.evaluate(x_train, y_train, verbose=0)
    _, test_acc = mlp_model.evaluate(x_test, y_test, verbose=0)

    print("train_acc: ", train_acc)
    print("test_acc: ", test_acc)

    [1] train_acc: 0.9827990531921387
    [2] test_acc:  0.8741965293884277

The accuracy will vary slightly between the different executions. The train accuracy
is nearly perfect, while the test accuracy is at 87%. These percentages apply when the
training input is 60 minutes and the prediction output would be for the next 720 minutes
(12 hours). Depending on how many epochs one uses and the size of the timesteps the results
vary as well.

Saving A Model To S3
--------------------

> The next step is to save the model to a AWS S3 Bucket so we can access the model from
> potentially another application. Since the data we want to predict will not be normalised

Reviewer: ```suggestion
The next step is to save the model to a AWS S3 Bucket so the model can be accessed from
other applications. Since the data we want to predict will not be normalised
```

> like the model, we also need to save the MinMaxScaler we created before alongside the

Reviewer: ```suggestion
like the model, we also need to save the ``MinMaxScaler`` we created before alongside the
```

TensorFlow model.

.. code-block:: python

    import os
    import boto3
    from joblib import dump

    bucket = "some-s3-bucket-name"

    model_name = "some-model-name" + ".h5"
    mlp_model.save(model_name)

    scaler_name = "some-scaler-name" + "_scaler.pkl"
    dump(min_max_scaler, scaler_name)

    client = boto3.client("s3")
    client.upload_file(model_name, bucket, model_name)
    client.upload_file(scaler_name, bucket, scaler_name)

    os.remove(model_name)
    os.remove(scaler_name)

> Now we should see our model and scaler in the according S3 bucket and we can continue
> with loading it from possibly a different application.

Reviewer: ```suggestion
The model and scaler will now be present in the S3 bucket, available for use from other
applications.
```

> Loading A Model From S3

Reviewer: I would maybe bundle this section into the Predicting section below, and have the code be both loading the S3 object and using it for predictions, to make it clear this is part of another application?

-----------------------

> Now that the model is saved in an AWS S3 Bucket, we can simply load it from there

Reviewer: ```suggestion
Now that the model is saved in an AWS S3 Bucket, we can load it from there
```

and use it to predict pump sensor data without having to first train a model.

.. code-block:: python

    bucket = "some-s3-bucket-name"

    model_name = "some-model-name"
    scaler_name = "some-scaler-name"

    client = boto3.client("s3")
    client.download_file(bucket, model_name, model_name)
    client.download_file(bucket, scaler_name, scaler_name)

Predicting With Pre-Trained Model
---------------------------------

> The models are now available, this means we can load them from file and start preparing
> the data to predict it. For simplicity we will take the same data that we used to

Reviewer: ```suggestion
Now that the application has the trained models, we can load them and use them
to make predictions from new sets of sensor data.
```

> train and test the model.

Reviewer: I think we can leave out the section about using the same data. We can fake that we're getting in new data from somewhere else, it's cooler that way. :wink:

.. code-block:: python

    import tensorflow as tf
    from joblib import load

    model = tf.keras.models.load_model(model_name)
    scaler = load(scaler_name)

    # We can use the df we queried in our very first step

>     new_data = df

Reviewer: I think that this section should be clear in that its a different application from before, so maybe loading this dataframe in from a diffrent source could emphasise that (or at least, loading it again from CrateDB).

    data_cols = df.iloc[:, 2:54]
    norm_data_cols = scaler.transform(data_cols)

    pred_pairs = []
    # the timestep is still the same as in the earlier part
    pred_saved_step = 0
    for i in norm_data_cols.index:
        row = pred_saved_step
        if row + timestep < len(df_train.index):
            data_window = norm_data_cols[row:row + timestep]
            data_vector = data_window.flatten()
            pred_pairs.append(data_vector)
            pred_saved_step = row + timestep
        else:
            break

    predictions = model.predict(pred_pairs.reshape(1, -1))
    predictions.flatten()

    [1] [0.8614458441734314, 0.8530051112174988, 0.8502672910690308, 0.8750132918357849, 0.8636448979377747, ...]

> Here we can se the first few predicted values, everything > 0.5 would mean that the machine is in a
> normal state, while < 0.5 would mean it is broken.

Reviewer: ```suggestion
We now have our predictions. Predictions > 0.5 mean that the machine is in a
normal state, while < 0.5 mean it is broken.
```
Do you know what prediction == 0.5 (unlikely as that is) would be? Should we say that everything ≥ 0.5 is normal, or is that inaccurate?

Saving The Predicted Values
---------------------------

In our final step we will insert the predictions from the previous step into
our CrateDB instance.

.. code-block:: python

    import time

    query = "INSERT INTO ml.pump_sensors_predictions (prediction, ts, model_name) VALUES (%s, %s, %s);"

    with connect(conn_str) as conn:
        with conn.cursor() as cur:
            print("Inserting data...")
            for entry in predictions:
                cur.execute(query, (float(entry), time.time(), model_name))

> With this, we have successfully used a Deep-Learning Algorithm, namely MLP in a distributed fashion
> and uploaded our results to CrateDB. We trained our model in a e.g. cloud-application while prediction
> the data on a different machine, decoupling the training and the predicting.

Reviewer: ```suggestion
With this, we have successfully used a Deep-Learning algorithm, namely a multilayer perceptron.
Using CrateDB to store raw data and model predictions, alongside using S3 for blob storage, we
have created a distributed architecture where applications can use various pieces of this Machine
Learning pipeline. The training and prediction stages are decoupled, and can be distributed
across different machines, contexts and scenarios.
```
